spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ExtendedV2Writes.scala

```diff
@@ -90,6 +93,14 @@ object ExtendedV2Writes extends Rule[LogicalPlan] with PredicateHelper {
       }
       val newQuery = ExtendedDistributionAndOrderingUtils.prepareQuery(write, query, conf)
       o.copy(write = Some(write), query = newQuery)
+
+    case rd @ ReplaceData(r: DataSourceV2Relation, query, _, None) =>
+      val rowSchema = StructType.fromAttributes(rd.dataInput)
```

Comment:
Is the purpose of overriding `dataInput` just to keep the original row schema?

Reply:
Correct. The idea is to report a row schema without metadata columns.

Reply:
Why not use `r.output`, the write relation's schema? That's what we do for normal writes isn't it? Should we just apply the `ResolveOutputRelation` rule?

Reply:
We can't apply the regular resolution as described [here](https://github.com/apache/iceberg/pull/3661/files#r767936935). Spark uses `query.schema` for regular writes.